### PR TITLE
Fix: Storage service fails to upload file with tags (#2724)

### PR DESCRIPTION
**Pull request by jessebrennan**
Author

- [x] PR title references issue
- [x] Title of main commit references issue
- [x] PR is linked to Zenhub issue

Author (reindex)

- [x] Added `r` tag to commit title                         <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                           <sub>or this PR does not require reindexing</sub>

Author (freebies & chains)

- [x] Freebies are blocked on this PR                       <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles              <sub>or there are no freebies in this PR</sub>
- [x] This PR is blocked by previous PR in chain            <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                <sub>or this PR is not chained to another PR</sub>

Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst  <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                         <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                           <sub>or this PR does not require upgrading</sub>

Author (requirements, before every review)

- [x] Ran `make requirements_update`                        <sub>or this PR leaves requirements*.txt, common.mk and Makefile untouched</sub>
- [x] Added `R` tag to commit title                         <sub>or this PR leaves requirements*.txt untouched</sub>
- [x] Added `reqs` label to PR                              <sub>or this PR leaves requirements*.txt untouched</sub>

Author (before every review)

- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`, squashed old fixups

Primary reviewer (before pushing merge commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Build passes in `sandbox`                             <sub>or added `no sandbox` label</sub>
- [x] Reindexed `sandbox`                                   <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit
- [x] Moved linked issue to Merged
- [ ] Pushed merge commit to Github

Primary reviewer (after pushing merge commit)

- [ ] Moved freebies to Merged column                       <sub>or there are no freebies in this PR</sub> 
- [ ] Shortened chain                                       <sub>or this PR is not the base of another PR</sub>
- [ ] Verified that `N reviews` labelling is accurate
- [ ] Commented on demo expectations                        <sub>or labeled as `no demo`</sub>
- [ ] Pushed merge commit to Gitlab                         <sub>or this changes can be pushed later, together with another PR</sub>
- [ ] Deleted PR branch from Github and Gitlab

Primary reviewer (reindex) 

- [ ] Started reindex in `dev`                              <sub>or this PR does not require reindexing `dev`</sub>
- [ ] Deleted BQ slot committment in `dev`                  <sub>or this PR does not require reindexing `dev`</sub>
- [ ] Checked for failures in `dev`                         <sub>or this PR does not require reindexing `dev`</sub>
- [ ] Started reindex in `prod`                             <sub>or this PR does not require reindexing `prod`</sub>
- [ ] Deleted BQ slot committment in `prod`                 <sub>or this PR does not require reindexing `prod`</sub>
- [ ] Checked for failures in `prod`                        <sub>or this PR does not require reindexing `prod`</sub>

Primary reviewer

- [ ] Unassign PR from reviewer and assign to PM

### CODE REVIEW

File: src/azul/service/storage_service.py [hunk 1]
```diff
@@ -82,9 +82,13 @@ def upload(self,
...
                                 Key=object_key,
-                                ExtraArgs=self._object_creation_kwargs(content_type, tagging))
+                                ExtraArgs=self._object_creation_kwargs(content_type))
+        # upload_file doesn't support tags so we need to make a separate request
+        # https://stackoverflow.com/a/56351011/7830612
+        if tagging:
+            self.put_object_tagging(object_key, tagging)
 
-    def _object_creation_kwargs(self, content_type, tagging):
+    def _object_creation_kwargs(self, content_type=None, tagging=None):
```
Comment (noah-aviel-dove) [inline src/azul/service/storage_service.py:91]: I think this would benefit from type annotations for the parameters.

Comment (noah-aviel-dove) [reply src/azul/service/storage_service.py:91]: And I suspect that the parameters ought to be made keyword-only now that they're optional, but I'm not certain.

**Review by noah-aviel-dove — APPROVED** [review]

File: src/azul/service/storage_service.py (hunk 1 above)
Comment (jessebrennan) [reply src/azul/service/storage_service.py:91]: I made both changes. I figure it can't hurt!

**Review by noah-aviel-dove — APPROVED** [review]
Looks good to me!

File: test/service/test_storage_service.py [hunk 2]
```diff
@@ -36,6 +37,21 @@ class StorageServiceTest(AzulUnitTestCase):
...
+        storage_service.create_bucket()
+
+        with tempfile.NamedTemporaryFile('w') as f:
+            f.write('some contents')
+            for tags in (None, {}, {'Name': 'foo', 'game': 'bar'}):
+                with self.subTest(tags=tags):
+                    storage_service.upload(file_path=f.name, object_key='test_file', tagging=tags)
+                    if tags is None:
+                        tags = {}
+                    self.assertEqual(tags, storage_service.get_object_tagging('test_file'))
```
Comment (hannes-ucsc) [inline test/service/test_storage_service.py:53]: 2nd mention of key. Extract variable.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
The titles of the two commits should have "Add test" and "Add fix" bodies.

Push separately again before next review, just to be sure.

**Review by hannes-ucsc — APPROVED** [review]

**Pull request merged**